Adding p-multigrid method

roeltielen24 opened the pull request:
Provides a new example file with a p and hp multigrid solver.

Since any p-refinement yields non-nested spaces, there are several options for the restriction and prolongation matrices.

filiatra committed:
add queries for gsMultiBasis (interior,boundary etc

filiatra committed:
add t as FunctionExpr parameter

filiatra committed:
added gsMultiBasis::partition

filiatra committed:
small fix

filiatra committed:
add gsVector::assign

stefantakacs committed:
Merge branch 'stable' into multibasis_queries

stefantakacs committed:
Merge branch 'stable' into multibasis_queries

stefantakacs committed:
ws fix

stefantakacs committed:
just to try the diffs

stefantakacs committed:
Merge branch 'stable' into branch 'multibasis_queries'

roeltielen committed:
bug fixewd in findFreeUncoupled

stefantakacs committed:
doxygen

stefantakacs committed:
Update gsDofMapper.h

roeltielen committed:
Merge remote-tracking branch 'origin/ImproveExpressions' into multibasis_queries

roeltielen committed:
Fixing detachment from multibasis_queries

roeltielen committed:
Latest version before coupling Xbraid

roeltielen committed:
Merge remote-tracking branch 'origin/stable' into rtielen

filiatra committed:
merge stable 01

filiatra committed:
minor change gsMultiBasis.hpp

filiatra committed:
fix conflicts

filiatra committed:
add accessors to incomplete lu

filiatra committed:
fix bug in visitor

filiatra committed:
add example

filiatra committed:
fixes for dirichlet - does not work yet

filiatra committed:
add bib R. Tielen

stefantakacs commented:
Hi, we should discuss these days how to base this better on existing multigrid code.

filiatra committed:
merge stable

filiatra committed:
revert part of the PR

filiatra committed:
Merge remote-tracking branch 'origin/stable' into rtielen

filiatra committed:
Merge remote-tracking branch 'origin/stable' into rtielen

stefantakacs committed:
Update gismo.bib

More standard bib file

stefantakacs committed:
Merge branch 'stable' into rtielen

stefantakacs committed:
minor

stefantakacs committed:
Could not find any indication that smoother 4 is actually implemented

stefantakacs committed:
Use consistently own classes for smoothing

stefantakacs committed:
simplify

stefantakacs committed:
simplify

stefantakacs committed:
these two branches are the same up to 1 char

simplify: typeCycle can only be 1 or 2 anyways

stefantakacs committed:
Seems to be not required

stefantakacs committed:
We do not want to pay for setup of coarse solver only once

stefantakacs committed:
might be handy later

stefantakacs committed:
whitespace to gismo standard

stefantakacs committed:
This has actually never been implemented (the implemented function has a
different signature).

stefantakacs committed:
what we actually do have is block ilut, not block gs

stefantakacs committed:
Why should one need more than one MG object?

stefantakacs committed:
fix

stefantakacs committed:
use for loop for this

stefantakacs committed:
Benchmarks >= 7 are broken

stefantakacs committed:
The solver does not need to know about the chosen benchmark problem.
Problem setup is done at one single place

stefantakacs committed:
Provide feedback here

stefantakacs committed:
simplify

stefantakacs committed:
remove unused parameters

stefantakacs committed:
No need for re-assembling of finest matrix

stefantakacs committed:
Remove unused parameters from setup and followup

Improve command line docs

stefantakacs committed:
Removing inherence allows to further simplify since members are not
passed as arguments

stefantakacs committed:
better validate input; rename numLevels to level where appropriate

stefantakacs committed:
if mg is used as a preconditioner, we now finally just call step, not
solve

This allows further simplifications

stefantakacs committed:
fix c15474e916d9585bbcb0764c7

stefantakacs committed:
unused

stefantakacs committed:
If one asks nice enough, the gsConjugateGradient class does what we want

stefantakacs committed:
whats that for?

stefantakacs committed:
Bicgstab can have its own class

stefantakacs committed:
Also the stand-alone case can be represented as an iterative solver.
This allows to combine all cases into a single one.

Moreover, do not claim convergence if wrong

stefantakacs committed:
proper use of RowMajor + simplifications

stefantakacs committed:
Avoid re-assembling the same matrix again and again and again

stefantakacs committed:
Use linear operators for restriction and prolongation

stefantakacs committed:
Remove debug data

stefantakacs committed:
Add bicgstab solver

Co-authored-by: roeltielen <roeltielen@hotmail.com>

stefantakacs committed:
simple unittest

stefantakacs committed:
make restart thereshold configurable

stefantakacs committed:
Merge branch 'bicgstab' into rtielen

stefantakacs committed:
the partition function is a total overkill if one only needs the number
of free uncoupled dofs

moreover, the function partition() is dubious since it does not ask for the additional information which is necessary to setup a dof mapper (like the interface strategy and the Dirichlet strategy). After setting up the dof mapper,
the function paritition() only accesses the dofmapper, so it might actually be a member of dofmapper instead.

stefantakacs committed:
Merge branch 'stable' into rtielen

stefantakacs committed:
fixes of use of eigen

stefantakacs committed:
Merge branch 'stable' into rtielen

stefantakacs committed:
slighly improve output

stefantakacs committed:
simplify block ilut, fix handling of multi-patch domains, simplify
signatures of functions for assembling transfer matrices

stefantakacs committed:
simplify the intergrid assemblers a bit

stefantakacs committed:
Merge branch 'stable' into rtielen

stefantakacs committed:
further simplify

hp should be a vector of indices

stefantakacs committed:
extend gsMultiGrid

provide the features that are necessary to behave like
pMultigrid_example

stefantakacs committed:
Finally, we can base everything on the gsMultiGrid class

stefantakacs committed:
In order not to pass around so long lists of parameters, the caller
should directly speak to the multigrid class

The tranfsfer time was not given; this is now done

There is no need to give the smoother setup time for smoothers which are
not actually used. So, the category "Smoother setup time" is sufficient.

stefantakacs committed:
further simplify

stefantakacs committed:
upd multigrid

stefantakacs committed:
Merge branch 'stable' into rtielen

stefantakacs committed:
fix warning

stefantakacs committed:
fix block ilut-smooter

how did this ever work without using the permutation?

stefantakacs committed:
these things are not required as arrays

stefantakacs committed:
simplify

stefantakacs committed:
turn block ilut into a linear operator since we can easily implement the
apply member

stefantakacs committed:
use lambda

stefantakacs committed:
further simplify

stefantakacs committed:
Merge branch 'stable' into rtielen

stefantakacs committed:
unified naming

stefantakacs committed:
the block ilut also needs boundary conditions+boundary handling to
obtain a proper dof mapper

simplify command line parameters; numPatches should actually numSplits
(or actually numPatchSplits) since in does not count the number of
patches

add some more cmd line parameters, like damping

stefantakacs committed:
Simplify assembling function: tell what they do and use them accordingly

stefantakacs committed:
simplify the assembling + galerin stuff such that it is more clear that
it is a if-then-else construct

Moreover, reduce the members a bit, so we do not need a vector of all
the assemblers.

Finally remove the assembler as template parameter since its not generic anyway.

stefantakacs committed:
basisH is obviously unused; rename basisL just to basis

stefantakacs committed:
include the setup in the main function and simplifications

stefantakacs committed:
cleanup the setup strategies a bit

provide a meaningful default for --Coarsening

bases do not need to be handled via pointers

stefantakacs committed:
provide docs

validate input and write all options (in one or an other way) back to the
console

stefantakacs commented:
@filiatra : Could you have a look on this PR?

stefantakacs committed:
minor changes

stefantakacs committed:
presentation

stefantakacs committed:
Merge branch 'stable' into rtielen

stefantakacs committed:
Merge pull request #527 from gismo/stable

upd p-multigrid
